bmp388/bmp388.go

```diff
@@ -0,0 +1,246 @@
+
+	return nil
+}
+
+// ReadTemperature returns the temperature in centicelsius, i.e 2426 / 100 = 24.26 C
+func (d *Device) ReadTemperature() (int32, error) {
+
+	rawTemp, err := d.readSensorData(REG_TEMP)
+	if err != nil {
+		return 0, err
+	}
+
+	// pulled from C driver: https://github.com/BoschSensortec/BMP3-Sensor-API/blob/master/bmp3.c
+	partialData1 := rawTemp - (256 * int64(d.cali.t1))
+	partialData2 := int64(d.cali.t2) * partialData1
+	partialData3 := (partialData1 * partialData1)
+	partialData4 := partialData3 * int64(d.cali.t3)
+	partialData5 := (partialData2 * 262144) + partialData4
+	partialData6 := partialData5 / 4294967296
+
+	d.cali.tlin = partialData6
+	temp := ((partialData6 * 25) / 16384)
+	return int32(temp), nil
+}
+
+// ReadPressure returns the pressure in centipascals, i.e 10132520 / 100 = 101325.20 Pa
+func (d *Device) ReadPressure() (int32, error) {
+
+	_, err := d.ReadTemperature()
+	if err != nil {
+		return 0, err
+	}
+	rawPress, err := d.readSensorData(REG_PRESS)
+	if err != nil {
+		return 0, err
+	}
+
+	// code pulled from bmp388 C driver: https://github.com/BoschSensortec/BMP3-Sensor-API/blob/master/bmp3.c
+	partialData1 := d.cali.tlin * d.cali.tlin
+	partialData2 := partialData1 / 64
+	partialData3 := (partialData2 * d.cali.tlin) / 256
+	partialData4 := (int64(d.cali.p8) * partialData3) / 32
+	partialData5 := (int64(d.cali.p7) * partialData1) * 16
+	partialData6 := (int64(d.cali.p6) * d.cali.tlin) * 4194304
+	offset := (int64(d.cali.p5) * 140737488355328) + partialData4 + partialData5 + partialData6
+	partialData2 = (int64(d.cali.p4) * partialData3) / 32
+	partialData4 = (int64(d.cali.p3) * partialData1) * 4
+	partialData5 = (int64(d.cali.p2) - 16384) * d.cali.tlin * 2097152
+	sensitivity := ((int64(d.cali.p1) - 16384) * 70368744177664) + partialData2 + partialData4 + partialData5
+	partialData1 = (sensitivity / 16777216) * rawPress
+	partialData2 = int64(d.cali.p10) * d.cali.tlin
+	partialData3 = partialData2 + (65536 * int64(d.cali.p9))
+	partialData4 = (partialData3 * rawPress) / 8192
+
+	// dividing by 10 followed by multiplying by 10
+	// To avoid overflow caused by (pressure * partial_data4)
+	partialData5 = (rawPress * (partialData4 / 10)) / 512
+	partialData5 = partialData5 * 10
+	partialData6 = (int64)(uint64(rawPress) * uint64(rawPress))
+	partialData2 = (int64(d.cali.p11) * partialData6) / 65536
+	partialData3 = (partialData2 * rawPress) / 128
+	partialData4 = (offset / 4) + partialData1 + partialData5 + partialData3
+	compPress := ((uint64(partialData4) * 25) / uint64(1099511627776))
+	return int32(compPress), nil
+}
+
+// ReadAltitude estimates the altitude above sea level in centimeters.
+func (d *Device) ReadAltitude() (alt int32, err error) {
+	press, err := d.ReadPressure()
+	if err != nil {
+		return
+	}
+	// pulled from bme280 driver, modified a bit to return altitude in centimeters
+	alt = int32(4433000.0 * (1.0 - math.Pow(float64(press)/float64(d.Config.SeaLevelPressure), 0.1903)))
+	return
+}
```

Comment:
I don't know what @deadprogram thinks about it but my personal opinion is that this should not be put in the driver itself. Instead, such related algorithms can be put in a separate package (maybe part of the drivers repository, maybe not) for miscellaneous calculations and conversions. Because while this is undoubtedly useful, it is relevant to all pressure sensors and not just this pressure sensor.

Reply:
I agree that that's a better way to do it. There's more than one way to estimate the altitude, so it can be left up to the end user on how they want to do it. I'll pull this out.